### PR TITLE
Handle cap limit with `maybeInt` type

### PR DESCRIPTION
**Pull request by tsujigiri**
Required by ad2games/gocake#13

### CODE REVIEW

**Comment by dziemba** [comment]
I was always wondering when we would run into a backwords-incompatible change...

This change breaks the wire format, so we can't do it like this, otherwise all old data will be unreadable. 

There's only two options IMO:
- We could give the new field a different field ID. That would mean that the old data will still be readable, but have the new field set to an empty value. We could also keep the old field and add logic that can read both field in the client.
- Instead of replacing the integer field, we add another boolean field to the top-level structure.
  That keeps old data intact, but we don't have the value for the boolean field then for old data, of course.

Not sure what's best solution. Probably the first option. WDYT?

**Comment by tsujigiri** [comment]
The second option didn't work because we seem to not be able to read a field from the XML twice. 🤔 In general, I'm in favor of the first option, because it is more explicit and you cannot forget to check if the value it returns was actually set by looking up a second value.

**Comment by dziemba** [comment]
Yea, but then the problem is how do you distinguish "has_value = false, because CAKE value was nil" from "has_value = false, because we have old data where the field is not set"? 

Or is that not a problem for our use case?

**Comment by tsujigiri** [comment]
The field is useless for our one use case without this information anyway and we don't need historical data ATM.

**Comment by dziemba** [comment]
Okay, then let's just give it a new ID.

**Comment by tsujigiri** [comment]
In general, shouldn't it be possible to use a field from the source-XML more than once, though?

**Comment by dziemba** [comment]
No, it's a limitation of the golang XML parser. See https://play.golang.org/p/szx8T5OBvl

**Comment by tsujigiri** [comment]
It seems like this limitation is an issue to us, because we are not separating the parsing and the transforming steps. If we were to parse the XML into an equivalent data structure and just access that in the transforming step, we could just use und re-use the data however we wanted to.

**Comment by dziemba** [comment]
In a broader sense, yes, I agree 👍 

I'm not very happy with the current design of gocake anymore, if I had to do it all over again, I'd probably get rid of all the "transform-structs" and `protocopy` and just have transform functions, something like this:

``` go
func (o *cakeproto.Offer) parseXMLNode(name string, node xml.Node) {
  switch (name) {
    case "id":
      o.Id = xmlToInt(node)
    case "name":
      o.Name = xmlToString(node)
    case "advertiser":
      o.Advertiser = &cakeproto.Advertiser{}
      o.Advertiser.parseXMLNode(node)
  }
}

```

That's almost as compact, is a lot faster because it doesn't use reflection, allows direct programming of custom cases without callback hell and gets rid of the ugly `protocopy` package. 

But since there was no real need for a refactoring, I never got around to do this.

**Comment by tsujigiri** [comment]
Still seems like a good idea. :smile:

**Comment by dziemba** [comment]
I'd also like to do it, but IMHO this isn't justifiable from a business perspective at the moment.

It sounds like a nice Hack Friday project, though ;)

**Comment by tsujigiri** [comment]
That's what I was thinking.

File: cap_states.proto [hunk 1]
```diff
@@ -21,6 +21,7 @@ message CapState {
   optional TimeWithZone start_date = 10;
   optional TimeWithZone end_date = 11;
   optional int64 current = 12;
-  optional int64 limit = 13;
+  optional int64 deprecated_limit = 13;
```
Comment (dziemba) [inline cap_states.proto]: Can you give this a better name? Right now it looks like a CapState has a "limit" and another value "deprecated limit". Maybe `limit_DEPRECATED`? Or just use `reserved 13;` (with comment)?

**Comment by dziemba** [comment]
🍄 

But since this is used by many projects, please add a detailed commit message explaining the change :)

**Comment by tsujigiri** [comment]
Will do. I was wondering, though, how to handle the deployment. The protocol is backwards compatible, but since we cannot keep filling the old and the new limit fields at the same time, the data isn't. We will have to deploy this change and the projects that depend on the limit, roughly at the same time, I guess. For the DWH, this is not so big of a deal. I'm not sure about the adserver, though. And I'd rather fix gocake instead, so we can keep filling both values for as long as we need to. What I don't want to do is go down another rabbit hole right now, only to fix the cap import in the dwh... Any thoughts on that?

**Comment by dziemba** [comment]
Well, it's quite easy now that we only added a new field, we just have to do it in this order:
1. Release new cakeproto version that has both the new and the old field (i.e. merge this PR)
2. Update all readers to support both fields (i.e. DWH + adserver)
3. Update all writers to write the new fields (i.e. gocake)
4. (Optional) Remove support for old field from all readers when no longer needed

If we stick to this, we shouldn't have any problems and also don't need to "deploy things at the same time".

**Comment by tsujigiri** [comment]
Yes. What I mean, though, is that this holds true for the protobuf definitions only, not for the actual data, since we cannot fill both fields at the same time.

**Comment by dziemba** [comment]
The idea is that you need to adapt all _readers_ to support _both_ versions, something like `limit = cap.limit || cap.DEPRECATED_limit`.  There is no other way, except for deploying everything at once and watching it go up in flames 😉

**Comment by tsujigiri** [comment]
Which works in this case, because the default value for the new type is nil. If it was an integer, for example, it would just return the default value. Ideally, I think, we would be able to fill both fields, then update the projects one after another, and eventually drop the old field, depending on whether we think we will ever need historical data. Which in this case we probably don't, since it's broken anyway.

**Comment by dziemba** [comment]
That would also work, of course. It's just a lot more work.

**Comment by tsujigiri** [comment]
Getting there, yes. Changing all the consumers once instead of twice, not so much. :wink:

**Comment by dziemba** [comment]
With either approach you only _have to_ change each project once. The rest is just cleaning up if you don't need support for the old data anymore.

**Comment by dziemba** [comment]
The advantage of my approach (adding support for both versions in the readers) is that you can support old data forever, if you need. And you can decide on a per-project basis if you need to.

**Pull request merged**